core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala

```diff
@@ -61,10 +64,12 @@ class ParamLiteral(_value: Any, _dataType: DataType, val pos: Int)
     31 * (31 * Objects.hashCode(dataType)) + Objects.hashCode(pos)
   }
 
+  var considerUnequal = false
+
   override def equals(obj: Any): Boolean = obj match {
     case a: AnyRef if this eq a => true
     case pl: ParamLiteral =>
-      pl.dataType == dataType && pl.pos == pos
+      pl.dataType == dataType && pl.pos == pos && !(pl.considerUnequal || considerUnequal)
```

Comment:
As explained somewhere above, this was added to enforce inequality between two paramliterals even though their position and datatype are same. And yes this was added for transforming the broadcast sub plan and reprepare the broadcast variables. Better would be to replace the paramliterlas with a new literal object itself, which I am doing for injecting new constants for UI plan display.. So I am removing this.